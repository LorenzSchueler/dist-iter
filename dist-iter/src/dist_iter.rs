use mpi::{
    topology::SimpleCommunicator,
    traits::{Communicator, Source},
};

use crate::task::Task;

pub struct DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    inner: I,
    init: bool,
    send_count: usize,
    recv_count: usize,
    world: &'w SimpleCommunicator,
}

impl<'w, I> Iterator for DistIter<'w, I>
where
    I: Iterator,
    I::Item: Task,
{
    type Item = <I::Item as Task>::OUT;

    fn next(&mut self) -> Option<Self::Item> {
        if !self.init {
            for dest in 1..self.world.size() {
                if let Some(task) = self.inner.next() {
                    task.send(self.world.process_at_rank(dest));
                    self.send_count += 1;
                }
            }
        }
        if self.recv_count < self.send_count {
            let tag = <I::Item as Task>::TAG;
            let (data, status) = self.world.any_process().receive_with_tag(tag);
            self.recv_count += 1;

            if let Some(task) = self.inner.next() {
                task.send(self.world.process_at_rank(status.source_rank()));
                self.send_count += 1;
            }
            Some(data)
        } else {
            None
        }
    }
}

pub trait MyIterExt {
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self>
    where
        Self: Iterator + Sized,
        Self::Item: Task;
}

impl<I> MyIterExt for I
where
    I: Iterator,
    I::Item: Task,
{
    fn into_dist_iter(self, world: &SimpleCommunicator) -> DistIter<Self> {
        DistIter {
            inner: self,
            init: false,
            send_count: 0,
            recv_count: 0,
            world,
        }
    }
}
