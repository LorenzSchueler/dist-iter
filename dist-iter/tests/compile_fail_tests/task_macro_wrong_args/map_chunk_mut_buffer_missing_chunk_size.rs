use dist_iter::{map_chunk_task, DistIterator};

fn main() {
    [1, 2, 3, 4, 5]
        .into_iter()
        .dist_map_chunk(map_chunk_task!(|buf: &mut [i32]| {
            for item in buf {
                *item += 1;
            }
        }));
}
