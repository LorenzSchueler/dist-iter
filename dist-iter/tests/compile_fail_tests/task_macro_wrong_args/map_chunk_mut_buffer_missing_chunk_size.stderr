error: no rules expected the token `|`
 --> tests/compile_fail_tests/task_macro_wrong_args/map_chunk_mut_buffer_missing_chunk_size.rs:6:41
  |
6 |         .dist_map_chunk(map_chunk_task!(|buf: &mut [i32]| {
  |                                         ^ no rules expected this token in macro call
  |
note: while trying to match `INPUT_CHUNK_SIZE`
 --> src/task.rs
  |
  |     (INPUT_CHUNK_SIZE = $IN:literal, OUTPUT_CHUNK_SIZE = $OUT:literal, |$closure_param:ident: impl Iterator<Item = $in:ty>| -> impl IntoI...
  |      ^^^^^^^^^^^^^^^^
